src/main/scala/piconot/internal/PicoWhile.scala

```diff
@@ -0,0 +1,120 @@
+package piconot.internal
+
+import scala.language.implicitConversions
+import picolib.maze.Maze
+import picolib.semantics._
+import scala.collection.mutable.MutableList
+import scalafx.application.JFXApp
+
+/**
+  * @author ben
+  *
+  * This design and implementation leaves a *lot* to be desired. But I wanted to
+  * give a sample solution that wasn't too innovative on syntax design and that
+  * contained a couple of implementation techniques that you might not have
+  * thought of at first.
+  *
+  * No, the name "Picobor" is not a typo -- this is a boring language :).
+  */
+
+class PicoWhile(val mazeFilename: String) extends JFXApp {
+
+  // the list of rules, which is built up as the Picobor program executes
+  private val rules = MutableList.empty[Rule]
+  private val program = PicoWhile.this
+
+  def addRule(rule: Rule) = rules += rule
+
+  def run() = {
+    val maze = Maze(mazeFilename)
+    val picobot =
+      new Picobot(maze, rules.toList) with TextDisplay with GUIDisplay
+    stage = picobot.mainStage
+    picobot.run()
+  }
+
+  /////////////////////////////////////////////////////////////////////////////
+  // Internal DSL definition
+  /////////////////////////////////////////////////////////////////////////////
+
+  private var nextState: Int = 0
+
+  implicit def intToState(i: Int): State = State(i.toString)
+  def anySurr: Surroundings = Surroundings(Anything, Anything, Anything, Anything)
+
+  // converts to surroundings with one open direction
+  def free(dir: MoveDirection): Surroundings = dir match {
+    case North => Surroundings(Open, Anything, Anything, Anything)
+    case East => Surroundings(Anything, Open, Anything, Anything)
+    case West => Surroundings(Anything, Anything, Open, Anything)
+    case South => Surroundings(Anything, Anything, Anything, Open)
+  }
+
+  // helper function - returns the opposite relative description of surrounding
+  def opp(r: RelativeDescription): RelativeDescription = r match {
+    case Open => Blocked
+    case Blocked => Open
+    case Anything => Anything
+  }
+
+  // returns the "opposite surroundings"
+  def opposite(surr: Surroundings): Surroundings = {
+    Surroundings(opp(surr.north), opp(surr.east), opp(surr.west), opp(surr.south))
+  }
+
+  // ensures that states flow correctly - recursive
+  def fixStates(rules: List[Rule], loopState: Int): List[Rule] = rules match {
```

Comment:
This function could merit more explanation as to what exactly needs fixing, and how you're fixing it.